smu/geometry/topology_from_geom.py

```diff
@@ -0,0 +1,174 @@
+    BondTopology that are consistent with that.
+    Note that `bond_topology` will be put in a canonical form.
+    Args:
+      bond_length_distribution:
+      bond_topology:
+      geometry:
+    Returns:
+      TopologyMatches
+  """
+  result = dataset_pb2.TopologyMatches()    # To be returned.
+  if len(bond_topology.atoms) == 1:
+    return result    # empty.
+
+  # Will be used when comparing perceived BondTopology's.
+  serialized_starting_form = bond_topology.SerializeToString()
+
+  utilities.canonical_bond_topology(bond_topology)
+  distances = utilities.distances(geometry)
+
+  # First join each Hydrogen to its nearest heavy atom, thereby
+  # creating a starting BondTopology from which all others can grow
+  starting_bond_topology = hydrogen_to_nearest_atom(bond_topology, distances)
+  if starting_bond_topology is None:
+    return result
+
+# heavy_atoms = [a for a in bond_topology.atoms if a != dataset_pb2.BondTopology.AtomType.ATOM_H]
+# heavy_atom_indices = indices_of_heavy_atoms(bond_topology)
+  heavy_atom_indices = [
+      i for i, t in enumerate(bond_topology.atoms) if t != dataset_pb2.BondTopology.AtomType.ATOM_H
+  ]
+
+  # For each atom pair, a list of possible bond types.
+  # Key is a tuple of the two atom numbers, value is an np.array
+  # with the score for each bond type.
+
+  bonds_to_scores: Dict[Tuple[int, int], np.array] = {}
+  for c in itertools.combinations(heavy_atom_indices, 2):    # All pairs.
+    i = c[0]
+    j = c[1]
+    dist = distances[i, j]
+    if dist > THRESHOLD:
+      continue
+    btypes = np.zeros(4, np.float32)
+    for btype in range(0, 4):
+      #    print(f"Looking for pdfs of {bond_topology.atoms[i]} {bond_topology.atoms[j]} type {btype} dist {dist}")
+      btypes[btype] = bond_lengths.pdf_length_given_type(bond_topology.atoms[i],
+                                                         bond_topology.atoms[j], btype, dist)
+
+    if np.count_nonzero(btypes) > 0:
+      bonds_to_scores[(i, j)] = btypes
+
+  if not bonds_to_scores:    # Seems unlikely.
+    return result
+
+# print(f"Mol with {len(bond_topology.atoms)} has {bonds_to_scores}")
+  mol = smu_molecule.SmuMolecule(starting_bond_topology, bonds_to_scores, matching_parameters)
+
+  search_space = mol.generate_search_state()
+  for s in itertools.product(*search_space):
+    bt = mol.place_bonds(list(s))
+    if not bt:
+      continue
+
+      continue
+    utilities.canonical_bond_topology(bt)
+    if utilities.same_bond_topology(bond_topology, bt):
+      bt.is_starting_topology = True
+    bt.smiles = smu_utils_lib.compute_smiles_for_bond_topology(bt,
+                                                               include_hs=True,
+                                                               labeled_atoms=True)
+    result.bond_topology.append(bt)
+
+  if len(result.bond_topology) > 1:
+    result.bond_topology.sort(key=lambda bt: bt.score, reverse=True)
+
+  return result
+
+
+class TopologyFromGeom(beam.DoFn):
```

Comment:
Not a big deal, but this doesn't need to be a DoFn. You can pass named arguments (e.g. bond_lengths) directly to functions in a beam.Map

Reply:
I went looking for solutions to getting the bonding info into the pipeline, and this was the first I found, and it seemed simple enough.

But hardly surprising, seems there are many ways of doing that... Will leave as is for now...